File: bindings/haskell/src/ProvenServers/Gameserver.hs
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
--
-- | Game Server types for the proven-servers ABI.
--
-- All tag values match the Idris2 ABI discriminants exactly.

module ProvenServers.Gameserver
  (
    SessionType(..)
  , sessionTypeToTag
  , sessionTypeFromTag
  , PlayerState(..)
  , playerStateToTag
  , playerStateFromTag
  , MatchState(..)
  , matchStateToTag
  , matchStateFromTag
  ) where

import Data.Word (Word8)

-- ---------------------------------------------------------------------------
-- SessionType
-- ---------------------------------------------------------------------------

-- | Game session types.
--
-- Tags 0-4 (5 constructors).
data SessionType
  = Lobby  -- ^ Lobby (tag 0).
  | Match  -- ^ Match (tag 1).
  | Practice  -- ^ Practice (tag 2).
  | Spectator  -- ^ Spectator (tag 3).
  | Tournament  -- ^ Tournament (tag 4).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'SessionType' to its ABI tag value.
sessionTypeToTag :: SessionType -> Word8
sessionTypeToTag = fromIntegral . fromEnum

-- | Decode a 'SessionType' from its ABI tag value.
sessionTypeFromTag :: Word8 -> Maybe SessionType
sessionTypeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: SessionType)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- PlayerState
-- ---------------------------------------------------------------------------

-- | Game player states.
--
-- Tags 0-5 (6 constructors).
data PlayerState
  = Idle  -- ^ Idle (tag 0).
  | Queuing  -- ^ Queuing (tag 1).
  | Loading  -- ^ Loading (tag 2).
  | Playing  -- ^ Playing (tag 3).
  | Spectating  -- ^ Spectating (tag 4).
  | Disconnected  -- ^ Disconnected (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'PlayerState' to its ABI tag value.
playerStateToTag :: PlayerState -> Word8
playerStateToTag = fromIntegral . fromEnum

-- | Decode a 'PlayerState' from its ABI tag value.
playerStateFromTag :: Word8 -> Maybe PlayerState
playerStateFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: PlayerState)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- MatchState
-- ---------------------------------------------------------------------------

-- | Game match states.
--
-- Tags 0-5 (6 constructors).
data MatchState
  = Waiting  -- ^ Waiting (tag 0).
  | Starting  -- ^ Starting (tag 1).
  | InProgress  -- ^ InProgress (tag 2).
  | Paused  -- ^ Paused (tag 3).
  | Ending  -- ^ Ending (tag 4).
  | Complete  -- ^ Complete (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'MatchState' to its ABI tag value.
matchStateToTag :: MatchState -> Word8
matchStateToTag = fromIntegral . fromEnum

-- | Decode a 'MatchState' from its ABI tag value.
matchStateFromTag :: Word8 -> Maybe MatchState
matchStateFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: MatchState)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

File: bindings/elixir/lib/proven_servers/gameserver.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Gameserver do
  @moduledoc """
  Game Server types for the proven-servers ABI.
  
  Formally verified game server types.
  Mirrors the Idris2 module `GameserverABI.Types`.
  
  - `SessionType` -- Game session types.
  - `PlayerState` -- Game player states.
  - `MatchState` -- Game match states.
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  # ===========================================================================
  # SessionType (tags 0-4)
  # ===========================================================================

  @typedoc """
  SessionType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type session_type :: :lobby | :match | :practice | :spectator | :tournament

  @session_type_tags %{
    lobby: 0,
    match: 1,
    practice: 2,
    spectator: 3,
    tournament: 4,
  }

  @tag_to_session_type Map.new(@session_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `SessionType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Gameserver.session_type_from_tag(0)
      {:ok, :lobby}
  """
  @spec session_type_from_tag(non_neg_integer()) :: {:ok, session_type()} | :error
  def session_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_session_type, tag)}
  end

  def session_type_from_tag(_tag), do: :error

  @doc """
  Encode a `SessionType` to the C-ABI tag value.
  """
  @spec session_type_to_tag(session_type()) :: non_neg_integer()
  def session_type_to_tag(val) when is_map_key(@session_type_tags, val) do
    Map.fetch!(@session_type_tags, val)
  end

  @doc """
  All `SessionType` variants in tag order.
  """
  @spec all_session_types() :: [session_type()]
  def all_session_types, do: [:lobby, :match, :practice, :spectator, :tournament]

  # ===========================================================================
  # PlayerState (tags 0-5)
  # ===========================================================================

  @typedoc """
  PlayerState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type player_state :: :idle | :queuing | :loading | :playing | :spectating | :disconnected

  @player_state_tags %{
    idle: 0,
    queuing: 1,
    loading: 2,
    playing: 3,
    spectating: 4,
    disconnected: 5,
  }

  @tag_to_player_state Map.new(@player_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `PlayerState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Gameserver.player_state_from_tag(0)
      {:ok, :idle}
  """
  @spec player_state_from_tag(non_neg_integer()) :: {:ok, player_state()} | :error
  def player_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_player_state, tag)}
  end

  def player_state_from_tag(_tag), do: :error

  @doc """
  Encode a `PlayerState` to the C-ABI tag value.
  """
  @spec player_state_to_tag(player_state()) :: non_neg_integer()
  def player_state_to_tag(val) when is_map_key(@player_state_tags, val) do
    Map.fetch!(@player_state_tags, val)
  end

  @doc """
  All `PlayerState` variants in tag order.
  """
  @spec all_player_states() :: [player_state()]
  def all_player_states, do: [:idle, :queuing, :loading, :playing, :spectating, :disconnected]

  # ===========================================================================
  # MatchState (tags 0-5)
  # ===========================================================================

  @typedoc """
  MatchState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type match_state :: :waiting | :starting | :in_progress | :paused | :ending | :complete

  @match_state_tags %{
    waiting: 0,
    starting: 1,
    in_progress: 2,
    paused: 3,
    ending: 4,
    complete: 5,
  }

  @tag_to_match_state Map.new(@match_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `MatchState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Gameserver.match_state_from_tag(0)
      {:ok, :waiting}
  """
  @spec match_state_from_tag(non_neg_integer()) :: {:ok, match_state()} | :error
  def match_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_match_state, tag)}
  end

  def match_state_from_tag(_tag), do: :error

  @doc """
  Encode a `MatchState` to the C-ABI tag value.
  """
  @spec match_state_to_tag(match_state()) :: non_neg_integer()
  def match_state_to_tag(val) when is_map_key(@match_state_tags, val) do
    Map.fetch!(@match_state_tags, val)
  end

  @doc """
  All `MatchState` variants in tag order.
  """
  @spec all_match_states() :: [match_state()]
  def all_match_states, do: [:waiting, :starting, :in_progress, :paused, :ending, :complete]

end
